### PR TITLE
Fixed a crash when inferring a value and using its qname on a slice

### LINKED ISSUE #8067: Inference crash: `Slice has no attribute 'qname'`
### Bug description

```python
x = slice(42)
x()
```

### Configuration

_No response_

### Command used

```shell
pylint asd.py
```

### Pylint output

```shell
************* Module asd
asd.py:1:0: C0114: Missing module docstring (missing-module-docstring)
Exception on node <Call l.2 at 0x1040c31f0> in file '/private/tmp/asd.py'
Traceback (most recent call last):
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 90, in walk
    callback(astroid)
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/method_args.py", line 72, in visit_call
    and inferred.qname() in self.linter.config.timeout_methods
AttributeError: 'Slice' object has no attribute 'qname'
asd.py:1:0: F0002: asd.py: Fatal error while checking 'asd.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/tusharsadhwani/Library/Caches/pylint/pylint-crash-2023-01-17-14-57-43.txt'. (astroid-error)
```

### Expected behavior

No crash.

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.2
Python 3.10.6 (main, Aug 30 2022, 04:58:14) [Clang 13.1.6 (clang-1316.0.21.2.5)]
```

### OS / Environment

MacOS Monterey

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #8067

### CODE REVIEW

**Review by DanielNoord — COMMENTED** [review]
I don't really like using `hasattr` as it is not exactly type safe. Can't we use `isinstance`?

**Comment by DanielNoord** [comment]
> Maybe we could check that it's a base.Instance ? I've seen that somewhere. Did you have something in mind ?

Purely based on the next line (`self.linter.config.timeout_methods`) I guess we're looking for `FunctionDefs` here?

**Comment by Pierre-Sassoulas** [comment]
Given our current test, it can be 
```
<class 'astroid.nodes.scoped_nodes.scoped_nodes.FunctionDef'>
<class 'astroid.nodes.scoped_nodes.scoped_nodes.ClassDef'>
<class 'astroid.bases.BoundMethod'>
```
(and create false negative according to our tests if we don't handle those cases). What's wrong with hasattr exactly ?

**Comment by DanielNoord** [comment]
> Given our current test, it can be
> 
> ```
> <class 'astroid.nodes.scoped_nodes.scoped_nodes.FunctionDef'>
> <class 'astroid.nodes.scoped_nodes.scoped_nodes.ClassDef'>
> <class 'astroid.bases.BoundMethod'>
> ```
> 
> (and create false negative according to our tests if we don't handle those cases). What's wrong with hasattr exactly ?

```python
class A:
    def qname(self):
        return 1
```

:smile: 

There is nothing that says that `qname` returns a string like we expect it to. I'm also not opposed to adding it to all `nodes` which seems much simpler than tackling these one by one.

**Comment by jacobtylerwalls** [comment]
> I'm also not opposed to adding it to all nodes which seems much simpler than tackling these one by one.

Marc was against that [here](https://github.com/PyCQA/astroid/pull/1284#pullrequestreview-823449216), but it also comes up regularly, see https://github.com/PyCQA/astroid/pull/1373.

**Comment by Pierre-Sassoulas** [comment]
Hmm yeah, right thank you I understand. The problematic version would be this:
```python
class A:
    def qname(self):
        return "threading.lock.acquire"
```
Because then it's a false positive for ``consider-using-with``. Are you proposing to add qname to all node in astroid directly ? I was going to propose that because clearly ``Slice`` has in fact a qname, and it's ``builtin.Slice``, right ?

**Comment by Pierre-Sassoulas** [comment]
Being able to use qname on all nodes would obviously be great in order to not crash when we get unexpected nodes. (Could assignAttr qname be the name of the variable with full namespace ?) But it's a big decisions with a required astroid upgrade, maybe hasattr is a decent temporary fix.

**Comment by DanielNoord** [comment]
Yeah I'm not sure. `hasattr` is just asking for trouble further down the line. When we add `py.typed` to `astroid` all of our incorrect calls to `qname()` for nodes that don't have it will be flagged anyway.
Here it does feel like we could get away with doing a check for `ClassDef`, `FunctionDef` and the "lowest type" of either `UnboundMethod` or `BoundMethod` (I always forget which inherits which).

**Comment by Pierre-Sassoulas** [comment]
Let's the how the primers react then

**Comment by Pierre-Sassoulas** [comment]
There's a whole lot of other call to hasattr accross the codebase (qname, attrname, name). Probably something to do in another refactor merge request or after the astroid qname discussion.

**Comment by tushar-deepsource** [comment]
Hey! This does fix the first crash in the issue, but it doesn't fix the second crash in `RefactoringChecker`, mentioned [here](https://github.com/PyCQA/pylint/issues/8067#issuecomment-1385095725).

Dropping the code snippet here as well:

```python
class Foo:
    def __init__(self, foo=slice(42)):
        self.foo = foo


def bar():
    i = Foo()
    i.foo()
    return 100
```

File: pylint/checkers/refactoring/refactoring_checker.py [hunk 1]
```diff
@@ -1998,7 +2000,7 @@ def _is_function_def_never_returning(self, node: nodes.FunctionDef) -> bool:
             )
         try:
             return node.qname() in self._never_returning_functions
```
Comment (DanielNoord) [inline pylint/checkers/refactoring/refactoring_checker.py:2002]: Clearly `node` is typed incorrectly, but oh well :smile:

**Review by DanielNoord — APPROVED** [review]

**Pull request merged**